### PR TITLE
Scal 90573 added test case for process trigger

### PR DESCRIPTION
**Pull request by gaurav-thoughtspot**
Task: https://thoughtspot.atlassian.net/browse/SCAL-90573

### CODE REVIEW

File: src/auth.spec.ts [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+    test('when session info giving error', async () => {
+        global.fetch = jest.fn(() => Promise.reject());
+        expect(await getSessionInfo(host)).toStrictEqual(mockSessionInfo);
+        global.fetch = window.fetch;
+    });
+
+    test('doTokenAuth: when authEndpoint and getAuthToken are not there, it throw error', async () => {
+        const embedConfig: any = {
+            thoughtSpotHost: host,
+            username: 'slgauravsharma',
```
Comment (kailash-choudhary-thoughtspot) [inline src/auth.spec.ts:51]: Can we use some username like tsuser

File: src/auth.spec.ts [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        global.fetch = window.fetch;
+        expect(loggedInStatus).toBe(true);
+    });
+
+    describe('doBasicAuth', () => {
+        const embedConfig: any = {
+            thoughtSpotHost: host,
+            username: 'slgauravsharma',
+            password: '12345678',
+        };
```
Comment (kailash-choudhary-thoughtspot) [inline src/auth.spec.ts:131]: Can we declare a common embed config for all four auth (success and failure)

File: src/auth.spec.ts [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+                    `${embedConfig.thoughtSpotHost}${EndPoints.BASIC_LOGIN}`
+                ) {
+                    return Promise.resolve({ status: 200 });
+                }
+                return Promise.reject();
+            });
+            await doBasicAuth(embedConfig);
+            expect(window.fetch).toBeCalled();
+            expect(loggedInStatus).toBe(true);
+            global.fetch = window.fetch;
```
Comment (kailash-choudhary-thoughtspot) [inline src/auth.spec.ts:161]: Can we shift this condition to before each, or try something like beforeEach(() => {
  fetch.mockClear();
});

**Review by kailash-choudhary-thoughtspot — DISMISSED** [review]

File: src/auth.spec.ts (hunk 1 above)
Comment (gaurav-thoughtspot) [reply src/auth.spec.ts]: updated

File: src/auth.spec.ts (hunk 2 above)
Comment (gaurav-thoughtspot) [reply src/auth.spec.ts]: updated

File: src/auth.spec.ts (hunk 3 above)
Comment (gaurav-thoughtspot) [reply src/auth.spec.ts]: updated

**Review by jbhanu-thoughtspot — APPROVED** [review]
lgtm

File: src/auth.spec.ts [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+    },
+    doTokenAuthFailureWithoutAuthEndPoint: {
+        thoughtSpotHost,
+        username: 'tsuser',
+        authEndpoint: '',
+        getAuthToken: null,
+    },
+    doTokenAuthFailureWithoutGetAuthToken: {
+        thoughtSpotHost,
+        username: 'tsuser',
```
Comment (jbhanu-thoughtspot) [inline src/auth.spec.ts:36]: let's just define variables at the top and re-use here.

const username = 'tsuser';
const password = '12345678';

doBasicAuth: {
        thoughtSpotHost,
        username,
        password,
    }

File: src/auth.spec.ts [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+describe('Unit test for auth', () => {
+    const originalWindow = window;
+    const mockSessionInfo = {
+        sessionId: '6588e7d9-710c-453e-a7b4-535fb3a8cbb2',
+        genNo: 3,
+        acSession: {
+            sessionId: 'cb202c48-b14b-4466-8a70-899ea666d46q',
+            genNo: 5,
+        },
+    };
```
Comment (jbhanu-thoughtspot) [inline src/auth.spec.ts:81]: These can be maintained above, no need to keep them inside describe.

File: src/auth.spec.ts [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        } catch (e) {
+            expect(e.message).toBe(
+                'Either auth endpoint or getAuthToken function must be provided',
+            );
+        }
+    });
+
+    test('doTokenAuth: when user is loggedIn', async () => {
+        global.fetch = jest.fn(async (param) => {
+            if (param === sessionInfoUrl) {
```
Comment (jbhanu-thoughtspot) [inline src/auth.spec.ts:124]: why do we have an if here? You would already know what type of value would be here, so just use the expected value.

File: src/auth.spec.ts [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+            }
+            return Promise.reject();
+        });
+        await doTokenAuth(embedConfig.doTokenAuthSuccess);
+        expect(loggedInStatus).toBe(true);
+    });
+
+    test('doTokenAuth: when user is not loggedIn & getAuthToken have response, isLoggedIn should called', async () => {
+        global.fetch = jest.fn(async (param, other) => {
+            if (param === sessionInfoUrl) {
```
Comment (jbhanu-thoughtspot) [inline src/auth.spec.ts:135]: let's not use if conditions inside a test case. If we have multiple use-cases, then they should be split into multiple tests, each testing that flow.

File: src/auth.spec.ts (hunk 6 above)
Comment (gaurav-thoughtspot) [reply src/auth.spec.ts:124]: In doTokenAuth have different API calls, we are mocking based on fetch URL based. and it's easy to test else we need to change the whole auth file import as mock and need to update all test cases based on spyon function-wise.

File: src/auth.spec.ts (hunk 7 above)
Comment (gaurav-thoughtspot) [reply src/auth.spec.ts:135]: In doTokenAuth have different API calls, we are mocking based on fetch URL based. and it's easy to test else we need to change the whole auth file import as mock and need to update all test cases based on spyon function-wise.

File: src/auth.spec.ts (hunk 5 above)
Comment (gaurav-thoughtspot) [reply src/auth.spec.ts]: Updated

File: src/auth.spec.ts (hunk 4 above)
Comment (gaurav-thoughtspot) [reply src/auth.spec.ts]: updated

**Review by jbhanu-thoughtspot — APPROVED** [review]
lgtm

**Pull request merged**